copy/copy.go

```diff
@@ -282,26 +354,127 @@ func (c *copier) copy(ctx context.Context, src, target string, overwriteTargetMe
 	return nil
 }
 
-func (c *copier) copyDirectory(ctx context.Context, src, dst string, stat os.FileInfo, overwriteTargetMetadata bool) (bool, error) {
+func (c *copier) include(path string, fi os.FileInfo) (bool, bool, error) {
+	matched := false
+	partial := true
+	for _, pattern := range c.includePatterns {
+		if fi.IsDir() {
+			pattern = strings.TrimSuffix(pattern, string(filepath.Separator))
+		}
+
+		if ok, p := prefix.Match(pattern, path); ok {
```

Comment:
Why do IncludePatterns/ExcludePatterns use different format/algorithm? Afaics only `ExcludePatterns` support `!`. Is there a difference of `ExcludePatterns` and potentially doing `!` in `IncludePatterns`.

Reply:
I wanted to match the logic in https://github.com/tonistiigi/fsutil/blob/8599091d69851a37058eec95da25c794ac060a0a/walker.go#L87

I think I'm following the same logic (it has separate handling for `IncludePatterns` and `ExcludePatterns` and doesn't appear to handle `!` in `IncludePatterns`), but maybe I'm missing something.

Reply:
I see. I don't remember why it is like this. I guess for `.dockerignore` is the only use-case atm that required `!` and that is the excludes list. `IncludePatterns` is not used at all anymore as well and `FollowPaths` is used instead. I guess the `FollowPaths` case where symlinks need to be followed does not apply here?

I think it is important that something like `COPY --filter` can be implemented with this. I'm not quite sure if atm client would need to convert some of the patterns to a negative version. But I guess if we would only use one input then with current implementation depending on if user wants to mostly define include or exclude paths one of them would be much slower. Matching with the options of walker is not the most important here so if you think there is a better/simpler logic we can use instead lmk.